File: my-folder/0383-ransom-note/solution.js
/**
 * @param {string} ransomNote
 * @param {string} magazine
 * @return {boolean}
 */
var canConstruct = function(ransomNote, magazine) {
    const map = new Map();

    for (c of magazine) {
        if (map.has(c)) {
            map.set(c, map.get(c) + 1);
        } else {
            map.set(c, 1);
        }
    }

    for (c of ransomNote) {
        if (!map.has(c)) return false;

        if (map.get(c) < 1) return false;

        map.set(c, map.get(c) - 1);
    }

    return true;
};

File: my-folder/0383-ransom-note/solution.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        rn_c = Counter(ransomNote)
        m_c = Counter(magazine)

        for k, v in rn_c.items():
            if k not in m_c or v > m_c[k]:
                return False
        
        return True
